contrib/registry/etcd/watcher.go

```diff
@@ -52,7 +53,16 @@ func (w *watcher) Next() ([]*registry.ServiceInstance, error) {
 	case <-w.ctx.Done():
 		return nil, w.ctx.Err()
 	case watchResp, ok := <-w.watchChan:
-		if !ok || watchResp.Err() != nil {
+		// Refer https://github.com/etcd-io/etcd/blob/5d45a88ab73cc5461ccf23e4df70598d8d433ff0/client/v3/watch.go#L65
+		if !ok && watchResp.Err() == nil {
+			// ctx is canceled or timed out
+			return nil, w.ctx.Err()
+		}
+		if watchResp.Err() != nil {
```

Comment:
What if there are other errors in handling this here?

Reply:
Do you want to recreate it as long as there are errors? If it's to solve the problem of idle time? gRPC will actively call to create a watch, and supports re creation.

Reply:
The `reWatch` logic was written by someone else, and I haven't modified it.

> Do you want to recreate it as long as there are errors? If it's to solve the problem of idle time? gRPC will actively call to create a watch, and supports re creation.

Reply:
Based on the documentation for `Watch`, I believe it may help resolve the `ErrCompacted` error.

> If revisions waiting to be sent over the watch are compacted, 
> then the watch will be canceled by the server, 
> the client will post a compacted error watch response, and the channel will close.